mobgap/Reorientation/CorrectSensorOrientationDynamic.py

```diff
@@ -0,0 +1,175 @@
+import pandas as pd
+import numpy as np
+from scipy.signal import correlate
+from sklearn.decomposition import PCA
+from gaitmap.trajectory_reconstruction import MadgwickAHRS
+from mobgap.Reorientation.utils import acceleration
+from mobgap.data_transform import (
+    SavgolFilter,
+    chain_transformers,
+)
+
+
+def CorrectSensorOrientationDynamic(data: pd.DataFrame, sampling_rate_hz: float) -> pd.DataFrame:
+
+    """
+    Corrects the orientation of the IMU data based on the orientation of the sensor.
+
+    Parameters
+    ----------
+    data
+        Accelerometer and gyroscope data. The data should be in m/s2 and deg/s.
+    sampling_rate_hz
+        Sampling rate of the data in Hz.
+
+    Returns
+    -------
+    IMU_corrected
+        Corrected IMU data.
+
+    Notes
+    -----
+    Points of deviation from the MATLAB implementation:
+    - The Madgwick filter is applied to the entire data array (walking bout) at once, as opposed to row by row in the
+    original implementation.
+    - The signal is already sliced before calling the present function (CorrectSensorOrientationDynamic).
+    Thus, this script has been simplified by removing a loop and a variable which were not needed.
+    - In MATLAB quaternions are represented as xyzw but python requires wxyz. Thus, the quaternions are adjusted.
+    - Parameters are expressed in the units used in mobgap,specifically, we use m/s^2 instead of g.
+    """
+
+    acc = data.iloc[:, 0:3]
+    gyr = data.iloc[:, 3:6]
+
+    # Madwick is called with the initial orientation specified inside the Matlab Madgwick function used in the Matlab
+    # implementation (UpdateIMUslt2). This choice was made since the first "initial orientation"
+    # specified in the Matlab implementation is overwritten in the UpdateIMUslt2 function.
+    
+    mad = MadgwickAHRS(beta=0.1, initial_orientation=np.array([-0.0863212587840360, -0.639984676042049, 
```

Comment:
Do we know where this initial orientation comes from? Does this initial orientation make sense?

Reply:
This is the initial orientation specified in the respective MATLAB function. 
To my understanding, the initial orientation describes a rotation where the human bends forwards slightly less that 90 degrees.